src/main/java/org/myrobotlab/service/Cron.java

```diff
@@ -81,29 +85,55 @@ public Task(Cron cron, String id, String cronPattern, String name, String method
 
     @Override
     public void run() {
-      if (cron != null) {
         log.info("{} Cron firing message {}->{}.{}", cron.getName(), name, method, data);
         cron.send(name, method, data);
-      } else {
-        log.error("cron service is null");
-      }
+        cron.history.add(new TaskHistory(id, new Date()));
+        if (cron.history.size() > cron.HISTORY_SIZE) {
+          cron.history.remove(0);
+        }
+        cron.broadcastState();
     }
 
     @Override
     public String toString() {
       return String.format("%s, %s, %s, %s", id, cronPattern, name, method);
     }
   }
+  
+  public static class TaskHistory {
+    public String id;
+    public Date processedTime;
+    
+    public TaskHistory(String id, Date now) {
+      this.id = id;
+      this.processedTime = now;
+    }    
+  }
 
   public final static Logger log = LoggerFactory.getLogger(Cron.class);
 
   private static final long serialVersionUID = 1L;
 
+  /**
+   * history buffer of tasks that have been executed
+   */
+  final protected List<TaskHistory> history = new ArrayList<>();
+
+  /**
+   * max size of history buffer
+   */
+  final int HISTORY_SIZE = 30;
+  
   /**
    * the thing that translates all the cron pattern values and implements actual tasks
    */
   transient private Scheduler scheduler = new Scheduler();
 
+  /**
+   * map of tasks organized by id
+   */
+  public Map<String, Task> tasks = new LinkedHashMap<>();

```

Comment:
Should probably be final, maybe not public

Reply:
oop :see_no_evil: